Push events to influxdb heapster sink

saad-ali opened the pull request:
This implements the second part of #118 (follow up to PR #192).

Related Kubernetes Issues:
* GoogleCloudPlatform/kubernetes/issues/4432
* GoogleCloudPlatform/kubernetes/issues/5638

Screenshot showing events in InfluxDB (pulled by heapster from Kubernetes API server and pushed by heapster to influxdb):
![influxdbevents](https://cloud.githubusercontent.com/assets/10052848/6913852/d412860a-d739-11e4-9b8e-5162b7f41370.png)

rjnagal commented:
ok to test

vishh commented:
Discussed with @saad-ali offline about the schema. He is working on updating the PR to use a more user friendly schema.

saad-ali commented:
PTAL. I refactored the decoder such that there are separate decoders for events and metrics--they create common `timeseries` objects that the drivers (gcm and influxdb) can consume.
And based on your feedback, events and metrics now share the common schema (I added "pod_name").

New screenshot showing events in InfluxDB (pulled by heapster from Kubernetes API server and pushed by heapster to influxdb):
![influxdb_events02](https://cloud.githubusercontent.com/assets/10052848/7015057/808f8c5a-dc82-11e4-904d-a41fa58a2805.png)

jimmidyson commented:
Can't we just store the marshalled event object as JSON rather than a string which is much harder to parse for display, etc?

vishh commented:
@saad-ali: I was thinking of just saving the event message as value. But it looks like you are storing a lot more data that that. If we think we need to store the entire event object, then I agree with @jimmidyson on marshalling the entire object into json. Sorry about the confusion.

saad-ali commented:
I think there's other information in there that is useful. I'll switch it to JSON.

jimmidyson commented:
Or we could create columns from the event data instead?

vishh commented:
@jimmidyson: Having columns derived from Event object is confusing. I feel that the event object wasn't conceived to be human parsable.

saad-ali commented:
@jimmidyson The original iteration of this PR was deriving columns from event objects. But after syncing with @vishh we agreed that this was contradictory to heapster's goals of having a common schema for all metrics/events. Hence this iteration.

jimmidyson commented:
Understand now about the common schema. Is this schema documented anywhere? Surely the storage schema might differ between sinks?

Storing marshalled as JSON should cover what anyone needs to do with it.

saad-ali commented:
@jimmidyson The idea was to try and maintain a common schema to the extent possible, and throw any additional information in to the value. Sink will always have the option to drop data that is irrelevant to them. The schema is documented here: [docs/storage-schema.md](https://github.com/GoogleCloudPlatform/heapster/blob/master/docs/storage-schema.md)

saad-ali commented:
PTAL.

New screenshot:
![influxdbevents05](https://cloud.githubusercontent.com/assets/10052848/7039487/6d13d384-dd77-11e4-9e96-824666f775b8.png)

vishh commented on sinks/api/types.go:
```diff
@@ -115,17 +123,14 @@ type MetricDescriptor struct {
 	Labels map[string]string
 
-	// The start and end time for which this data is representative.
+	// The start and end time for which this data is representative and the precision of the timestamps.
```
Precision is not included here.

saad-ali replied:
Fixed.

vishh commented on sinks/gcm/driver.go:
```diff
@@ -141,43 +151,52 @@ type metricWriteRequest struct {
-		if _, ok := self.exportedMetrics[metric.Name]; !ok {
-			return fmt.Errorf("unable to push unknown metric %q", metric.Name)
+	for _, inputTimeseriesList := range inputTimeseriesLists {
+		if inputTimeseriesList != nil {
```
nit: How about reversing the condition and continuing if its nil?

saad-ali replied:
Changed.

vishh commented on sinks/gcm/driver.go:
```diff
@@ -141,43 +151,52 @@ type metricWriteRequest struct {
-		if _, ok := metric.Value.(int64); !ok {
-			return fmt.Errorf("non-int64 data not implemented. Seen for metric %q", metric.Name)
+	for _, inputTimeseriesList := range inputTimeseriesLists {
+		if inputTimeseriesList != nil {
```
Ditto as above.

saad-ali replied:
Changed.

vishh replied:
I found the same pattern in a few other files. Can you make the same change in other files as well?

vishh commented on sinks/influxdb/driver.go:
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
+	for _, timeseriesList := range timeseriesLists {
+		if timeseriesList != nil {
+			for _, timeseries := range *timeseriesList {
+				if timeseries.TimePrecision == precision {
```
Why do we need this check? Why can't we use the precision that is passed via `timeseries.TimePrecision`?

saad-ali replied:
It is required for filtering out other precisions. Each Timeseries can specify it's own precision and the influxDB client expects all points in a WriteSeriesWithTimePrecision call to have the same precision. So we filter the timeseries, and issue one WriteSeriesWithTimePrecision call per precision type (if that type is present).

vishh replied:
Ah! Why do we care about precision at an API level? Why not pass timestamps with max precision and round to a sane precision in the sinks?

vishh commented on sinks/influxdb/driver.go:
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
+			sink.recordWriteFailure()
+			return err
+		}
+		glog.V(1).Info("Succesfully wrote %q timeseries to influxDB", precision)
```
This log file is misleading. Why are we logging precision in place of number of timeseries?

saad-ali replied:
See above comment about precision, essentially `storeTimeseriesWithPrecision` is called for each possible precision type (second, milli, micro). It filters any timeseries that match that precision, and issues a `WriteSeriesWithTimePrecision` for those points with that precision. So this log message is essentially saying that a time series with a given precision was written.

vishh replied:
Hmm. Maybe rephrase to `Successfully wrote timeseries to InfluxDB with precision: %q`?

vishh commented on sinks/api/types.go:
```diff
@@ -148,24 +153,35 @@ type SupportedStatMetric struct {
+	TimePrecision TimePrecision
+
+	// The labels used by this series.
+	LabelKeys []string
```
Is this for optimization? Why can't we use all the labels defined as part of the point?

saad-ali replied:
Yes, it is an optimization. Each `timeseries` can have multiple points, and any given point may specify some subset of labels for the timeseries. So, instead of iterating through all points and their lables to figure out the complete set of label keys, we have `LabelKeys` defined at the top defining all the labels that points in this `timeseries` may have (if any individual point defines a label that is not part of `labelKeys`, it is ignored, if any individual point does not provide a value for a label that is part of `labelKeys`, it is left empty).

vishh commented on sinks/influxdb/driver.go:
```diff
@@ -37,72 +37,130 @@ type influxdbSink struct {
+	columns := []string{}
+	columns = append(columns, "time")            // Column 0 header
+	columns = append(columns, "value")           // Column 1 header
+	columns = append(columns, "sequence_number") // Column 2 header
```
How about sharing setting up of these standard columns between the two methods here?

saad-ali replied:
Done.

vishh commented on sinks/api/eventdecoder.go:
```diff
@@ -0,0 +1,85 @@
+			}
+			points = append(points, Point{
+				Labels: getEventLabels(&event),
+				Start:  event.LastTimestamp.Time,
```
Should we round the timestamp here to a millisecond?

saad-ali replied:
Done.

vishh commented on sinks/api/metricdecoder.go:
```diff
@@ -121,24 +123,36 @@ func (self *defaultDecoder) getContainerMetrics(container *source_api.Container,
-				for _, point := range points {
+				seriesName := supported.Name
+				if supported.Units.String() != "" {
+					seriesName = fmt.Sprintf("%s_%s", seriesName, supported.Units.String())
```
I don't think GCM will like diverging from the registered series names. What ever series name we used while registering the events is what we should use here.
If Influxdb decides to mangle the series name with additional metadata, then it must be local to Influxdb.

vishh commented:
@saad-ali: The PR looks good. Just a few comments. Some of the comments are related to sink functionalities and unfortunately they are not documented properly :(

vishh commented on sinks/api/metricdecoder.go:
```diff
@@ -121,24 +123,36 @@ func (self *defaultDecoder) getContainerMetrics(container *source_api.Container,
 					labels := copyLabels(labels)
-					for name, value := range point.labels {
+					for name, value := range internalPoint.labels {
 						labels[name] = value
```
These labels might not show up as part of LabelKeys. Why not include all metric labels by default as part of labelkeys?

saad-ali commented:
Closing this. #234 replaces this.
